### PR TITLE
[RFC] Improved media modal

### PR DESCRIPTION
**Pull request by yuntan**
This pr is mainly for mobile devices.

- Pinch zoom and pan
- Larger image
- Larger close button (Solves #3660, #4743)
- ~~Close the modal by swiping vertically (suggested in #4743)~~
- Show/hide close button and right/left navigation on tapping image
- More blank space to close the modal
- ~~Zoom/reset zoom on double tap~~

Before | After
--|--
Screenshot of before|Screenshot of after

screencast (click to open YouTube):
[YouTube thumbnail](https://youtu.be/xJoJi8rAAwQ)

# TODOs (help and advice wanted!)
- [x] What the name of `ImageLoader` should be?
  + Separated zoom functionary to `ZoomableImage` component
- [x] What is `--amorphous` state of the image?
  + Amorphous means the size of image is not determined. Dealed with it (debugged by React dev tool).
- [ ] ~~Can't zoom on the tapping point on double tap~~
- [x] Got extra margin on the bottom of the modal (I can't find out why this happens).
  + Fixed.
- [ ] ~~The modal closes after some seconds delay from swiping.~~
- [ ] ~~What is the proper value of `VERTICAL_SWIPE_THRESHOLD_RATIO` ?~~
- [ ] ~~Better animation for vertical swiping~~
- [x] Not tested for video player
  + Tested and fixed.
- [x] Adjust margin of close button
  + Done.
- [x] What is the proper value of `MAX_SCALE` ~~and `DOUBLE_TAP_SCALE`~~ ?
  + `MAX_SCALE = 4` is enough.
- [ ] Not tested on iOS.
  + **Still help wanted!**

“cat” by waferboard is licensed under CC BY 2.0
“The Black Cat / Черная кошка” by Julay Cat is licensed under CC BY 2.0

### CODE REVIEW

**Comment by yuntan** [comment]
rebased to master

**Comment by yuntan** [comment]
I'll put "swipe to close" and "double tap to zoom" functionary to another pr.

**Comment by yuntan** [comment]
Almost done. Changed status to request for comment. I understand this PR is a bit large and takes some work to review. However, this pr can improve the mobile experience of mastodon dramatically (eg. you can close-up cute cut's face in a photograph of a cat and it's owner).

I've tested it on Chrome for PC and Chrome for Android. Can anyone test this pr on iOS?

File: app/javascript/mastodon/components/extended_video_player.js [hunk 1]
```diff
@@ -31,6 +32,12 @@ export default class ExtendedVideoPlayer extends React.PureComponent {
     this.video = c;
   }
 
+  handleClick = ev => {
```
Comment (akihikodaki) [inline app/javascript/mastodon/components/extended_video_player.js]: `ev` to represent event is not used in Mastodon. `e` or `event` is more common and would be preferred.
Anyway that is a trivial difference in style and can be ignored.

File: app/javascript/mastodon/features/ui/components/media_modal.js [hunk 2]
```diff
@@ -68,14 +70,21 @@ export default class MediaModal extends ImmutablePureComponent {
     return this.state.index !== null ? this.state.index : this.props.index;
   }
 
+  switchNavigation = () => {
```
Comment (akihikodaki) [inline app/javascript/mastodon/features/ui/components/media_modal.js]: maybe `toggleNavigation`

File: app/javascript/mastodon/features/ui/components/media_modal.js [hunk 3]
```diff
@@ -104,21 +134,40 @@ export default class MediaModal extends ImmutablePureComponent {
       alignItems: 'center', // center vertically
     };
 
+    const navigationClassName = classNames('media-modal__navigation', {
+      'media-modal__navigation--hidden': !navigationShown,
```
Comment (akihikodaki) [inline app/javascript/mastodon/features/ui/components/media_modal.js]: If CSS style uses `media-modal__navigation--hidden`, why don't you keep the state as `navigationHidden`?

File: app/javascript/mastodon/features/ui/components/media_modal.js [hunk 4]
```diff
@@ -104,21 +134,40 @@ export default class MediaModal extends ImmutablePureComponent {
...
-          </ReactSwipeableViews>
+        <div
+          className='media-modal__closer'
+          role='presentation'
+          onClick={onClose}
+        >
+          <div className='media-modal__content'>
+            <ReactSwipeableViews
+              style={{
+                // on Android Chrome, pixel length of 100vh is mutable
```
Comment (akihikodaki) [inline app/javascript/mastodon/features/ui/components/media_modal.js]: In what kind of situation `100vh` changes?

File: app/javascript/mastodon/features/ui/components/zoomable_image.js [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+
+const getMidpoint = (p1, p2) => ({
+  x: (p1.clientX + p2.clientX) / 2,
+  y: (p1.clientY + p2.clientY) / 2,
+});
+
+const getDistance = (p1, p2) =>
+  Math.sqrt(Math.pow(p1.clientX - p2.clientX, 2) + Math.pow(p1.clientY - p2.clientY, 2));
+
+const between = (min, max, value) => Math.min(max, Math.max(min, value));
```
Comment (akihikodaki) [inline app/javascript/mastodon/features/ui/components/zoomable_image.js]: What about to name `clamp`?

File: app/javascript/mastodon/features/ui/components/zoomable_image.js [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+  }
+
+  removers = [];
+  container = null;
+  image = null;
+  lastTouchEndTime = 0;
+  lastDistance = 0;
+
+  componentDidMount () {
+    let handler = this.handleTouchStart.bind(this);
```
Comment (akihikodaki) [inline app/javascript/mastodon/features/ui/components/zoomable_image.js]: You don't have to `bind` it.

File: app/javascript/mastodon/features/ui/components/zoomable_image.js [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+  handleTouchStart = ev => {
+    if (ev.touches.length !== 2) return;
+
+    this.lastDistance = getDistance(...ev.touches);
+  }
+
+  handleTouchMove = ev => {
+    const { scrollTop, scrollHeight, clientHeight } = this.container;
+    if (ev.touches.length === 1
+      && scrollTop !== scrollHeight - clientHeight) {
```
Comment (akihikodaki) [inline app/javascript/mastodon/features/ui/components/zoomable_image.js]: `&&` is placed in the earlier line in Mastodon's code.

File: app/javascript/mastodon/features/ui/components/zoomable_image.js [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+    // math memo:
+    // x = (scrollLeft + midpoint.x) / scrollWidth
+    // x' = (nextScrollLeft + midpoint.x) / nextScrollWidth
+    // scrollWidth = clientWidth * scale
+    // scrollWidth' = clientWidth * nextScale
+    // Solve x = x' for nextScrollLeft
+    const nextScrollLeft = (scrollLeft + midpoint.x) * nextScale / scale - midpoint.x;
+    const nextScrollTop = (scrollTop + midpoint.y) * nextScale / scale - midpoint.y;
+
+    // this.container.scrollLeft = nextScrollLeft;
```
Comment (akihikodaki) [inline app/javascript/mastodon/features/ui/components/zoomable_image.js]: To be removed?

File: app/javascript/mastodon/features/ui/components/zoomable_image.js [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+    // this.container.scrollTop = nextScrollTop;
+
+    this.setState({ scale: nextScale }, () => {
+      // callback
```
Comment (akihikodaki) [inline app/javascript/mastodon/features/ui/components/zoomable_image.js]: Anybody should be sure that and there is no need for the comment.

File: app/javascript/mastodon/features/ui/components/zoomable_image.js [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+    ev.stopPropagation();
+    const handler = this.props.onClick;
+    if (handler) handler();
+  }
+
+  setContainerRef = c => {
+    this.container = c;
+  }
+
+  setImageRef = c => {
```
Comment (akihikodaki) [inline app/javascript/mastodon/features/ui/components/zoomable_image.js]: What does `c` stand for?

**Review by akihikodaki — CHANGES_REQUESTED** [review]
I, who is a user of the mobile interface, strongly agree that this kind of feature is necessary.

I made lots of comments, but many of them are just to say "it is not Mastodon-way as far as I know," and trivial. At least, please:
* provide a detailed explanation for the comment saying "on Android Chrome, pixel length of 100vh is mutable". * remove redundant `bind`s.

File: app/javascript/mastodon/features/ui/components/zoomable_image.js (hunk 10 above)
Comment (Gargron) [reply app/javascript/mastodon/features/ui/components/zoomable_image.js]: `c` stands for component, we use this naming in `setRef` handlers commonly

**Comment by yuntan** [comment]
Thank you for @akihikodaki 's detailed indications. I've fixed designated parts of the code.

**Review by akihikodaki — APPROVED** [review]

**Comment by Gargron** [comment]
@akihikodaki Do you think this should get into the RC?

**Comment by akihikodaki** [comment]
@Gargron No. In my opinion, RC is to suspend all feature additions.

**Comment by Gargron** [comment]
I've read over the code... and given I already merged some features since rc1, I'm gonna take a chance on this, because it should be a really nice change.

**Comment by yuntan** [comment]
Thank you, @akihikodaki and @Gargron !

**Comment by penduin** [comment]
Does this finally allow images to be closed by using the "back" button?  Or does that still keep the image open and navigate away underneath?

**Pull request merged**